File: src/main/java/demo/dev/MockRedisServer.java
package demo.dev;

import java.io.IOException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import redis.embedded.RedisServer;

public class MockRedisServer {
    private static Logger LOG = LoggerFactory.getLogger(MockRedisServer.class);

    private final int port;
    private RedisServer redisServer = null;

    public MockRedisServer(int port) {
        this.port = port;
    }

    public void start() {
        try {
            redisServer = new RedisServer(port);
            if (!redisServer.isActive()) {
                redisServer.start();
                LOG.info("Mock Redis start");
            }
        } catch (IOException e) {
            throw new RuntimeException("error", e);
        }
    }

    public void stop() {
        if (redisServer != null) {
            LOG.info("Redis stop");
            redisServer.stop();
            redisServer = null;
        }
    }

    public int getPort() {
        return port;
    }
}

File: src/main/scala/demo/dev/MockRedisServer.scala
package demo.dev

import demo.ext.Logable
import redis.embedded.RedisServer
import demo.utils.Utils

trait MockRedisServerable {
  def start(): Unit

  def stop(): Unit
}

class MockRedisServer(val port: Int) extends MockRedisServerable with Logable {

  val redisServer = new RedisServer(port)

  override def start(): Unit = {
    if (!Utils.availablePort(port)) {
      logger.warn("端口{}已经被占用！", port)
      return
    } else if (!redisServer.isActive) {
      redisServer.start()
      logger.info("Mock Redis start")

      Runtime.getRuntime.addShutdownHook(new Thread() {
        override def run(): Unit = {
          if (redisServer.isActive) {
            redisServer.stop()
          }
        }
      })
    }
  }

  override def stop(): Unit = {
    if (redisServer.isActive) {
      redisServer.stop()
      logger.info("Mock Redis stop")
    }
  }
}